src/main/java/com/minecolonies/coremod/event/ClientEventHandler.java

```diff
@@ -367,29 +392,21 @@ private static void handleHutBlockResearchUnlocks(final IColony colony, final Li
     }
 
     /**
-     * Renders building bounding boxes into the client
-     *
-     * @param event  The caught event
-     * @param world  The world in which to render
-     * @param player The player for which to render
+     * Updates the blueprint cache by adding/removing buildings depending on their distance from the active building
+     * @param colony The colony which contains the buildings in the blueprint cache
+     * @param world The world in which to render
      */
-    private static void handleRenderBuildTool(@NotNull final RenderWorldLastEvent event, final ClientWorld world, final PlayerEntity player)
+    private static void updateBlueprintCache(final IColonyView colony, final ClientWorld world)
     {
-        if (Settings.instance.getActiveStructure() == null)
-        {
-            return;
-        }
-
-        final IColonyView colony = IColonyManager.getInstance().getClosestColonyView(world, new BlockPos(player.position()));
-        if (colony == null)
-        {
-            return;
-        }
-
-        final BlockPos activePosition = Settings.instance.getPosition();
-        final Map<BlockPos, Triple<Blueprint, BlockPos, BlockPos>> newCache = new HashMap<>();
+        final Map<BlockPos, Tuple<Blueprint, AxisAlignedBB>> newCache = new HashMap<>();
+
+        AxisAlignedBB activeBuildingBB = getBlueprintBoundingBoxes(activeBuildingBlueprint, Settings.instance.getPosition());
+        activeBuildingBB = activeBuildingBB.inflate(1);
+        // todo: optimise: This causes a big lag spike for the client when they open the build tool for the first time
+        //  after joining a world because it's loading something (pretty sure it's from loading all the colony's
+        //  building's blueprints at once). If we could distribute the work load over several frames, it should fix the spike.
+        // Update the building cache by adding buildings that are near the active building and removing buildings that are no longer near
         for (final IBuildingView buildingView : colony.getBuildings())
-
         {
             if (MinecoloniesAPIProxy.getInstance().getConfig().getClient().neighborbuildingrendering.get())
```

Comment:
Buidlings already have corner positions stored in the building tileentity and building, I'm sure we can leverage that and don't have to load the actual blueprint.